ecommerce/models_test.py

```diff
@@ -253,11 +253,12 @@ def test_product_version_save_text_id_badproduct(mocker):
     """ProductVersion.text_id should None if ProductVersion.product is invalid"""
     mock_log = mocker.patch("ecommerce.models.log")
     product_version = ProductVersionFactory.create(
-        product=ProductFactory.create(content_object=LineFactory())
+        product=ProductFactory.create(content_object=LineFactory()), id=1
```

Comment:
After upgrading, the tests were failing locally when switching to `assert_called_once_with` as [`called_once_with` is flagged as an error in 3.12.](https://github.com/miguelgrinberg/python-socketio/issues/1265#issuecomment-1784134829).

```
AssertionError: expected call not found.
E       Expected: error('The content object for this ProductVersion (2092) does not have a `text_id` property')
E         Actual: error('The content object for this ProductVersion (None) does not have a `text_id` property')
```

I'm not sure how this was working before, as it ideally shouldn’t work when creating a new ProductVersion. In the [log.error](https://github.com/mitodl/mitxpro/blob/master/ecommerce/models.py#L235), `self.id` is referenced before the ProductVersion is saved (i.e., `super().save(*args, **kwargs)`), meaning `self.id` should always be `None` in the "Actual" error before the new object is saved, which aligns with reality. This was added with that context in mind.

I was curious about there should be error traces if that's case. I’ve noticed similar past errors in [notifications-xpro-sentry slack](https://mitodl.slack.com/archives/G01LMMQTPL3/p1677016478203079) when adding a new product version from Django admin.

Reply:
Further updates specifically on this:

`mock_log.called_once_with() ` does nothing meaningful in the context of an assert statement because it's just checking if mock_log exists, which it always does.

**Why It Didn't Fail:**
If we write the test as assert mock_log.called_once_with(), Python interprets this as just asserting that mock_log (the mock object itself) exists, which it does. This kind of assertion would never fail unless mock_log was somehow None or False.

E.g:

```
(Pdb++) mock_log.called_once_with(f"The content object for this ProductVersion ({product_version.id}) does not have a `text_id` property")
<MagicMock name='log.called_once_with()' id='281473488743728'>
(Pdb++) 

(Pdb++) assert mock_log.called_once_with(f"The content object for this ProductVersion ({product_version.id}) does not have a `text_id` property")
(Pdb++) 
```

This passes because it's just checking that mock_log exists, which it does, but it doesn't check the actual function call and arguments.